src/main/java/org/openlmis/integration/dhis2/service/AuthService.java

```diff
@@ -0,0 +1,84 @@
+ * Copyright © 2017 VillageReach
+ *
+ * This program is free software: you can redistribute it and/or modify it under the terms
+ * of the GNU Affero General Public License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details. You should have received a copy of
+ * the GNU Affero General Public License along with this program. If not, see
+ * http://www.gnu.org/licenses.  For additional information contact info@OpenLMIS.org.
+ */
+
+package org.openlmis.integration.dhis2.service;
+
+import java.net.URI;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.commons.codec.binary.Base64;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestOperations;
+import org.springframework.web.client.RestTemplate;
+import org.springframework.web.util.UriComponentsBuilder;
+
+@Service
+public class AuthService {
+
+  static final String API_TOKEN_URL = "/api/apiToken";
+  static final String RESPONSE = "response";
+  static final String KEY = "key";
+
+  private final RestOperations restTemplate;
+
+  public AuthService() {
+    this(new RestTemplate());
+  }
+
+  public AuthService(RestTemplate restTemplate) {
+    this.restTemplate = restTemplate;
+  }
+
+  /**
+   * Retrieves access token from the auth service.
+   *
+   * @param username  Name of the specific user.
+   * @param password  User password.
+   * @param serverUrl User's server address.
+   * @return token.
+   */
+  public String obtainAccessToken(String username, String password, String serverUrl) {
+
+    String plainCreds = username + ":" + password;
+    byte[] plainCredsBytes = plainCreds.getBytes();
+    byte[] base64CredsBytes = Base64.encodeBase64(plainCredsBytes);
+    String base64Creds = new String(base64CredsBytes);
+
+    HttpHeaders headers = new HttpHeaders();
+    headers.add(HttpHeaders.AUTHORIZATION, "Basic " + base64Creds);
+
+    Map<String, String> emptyRequestBody = new HashMap<>();
+
+    HttpEntity<?> request = new HttpEntity<>(emptyRequestBody, headers);
+
+    UriComponentsBuilder uriBuilder = UriComponentsBuilder.newInstance().uri(URI.create(
+            serverUrl + API_TOKEN_URL
+    ));
+    URI uri = uriBuilder.build(true).toUri();
+
+    ResponseEntity<?> response = restTemplate.exchange(
+            uri,
+            HttpMethod.POST,
+            request,
+            Object.class);
+
+    return ((Map<String, Map<String, String>>) response.getBody()).get(RESPONSE).get(KEY);
```

Comment:
It would be useful to handle errors here.
For example, what if server is out of reach (offline/wrong url) - is there Spring exception?
What if response is not as expected - that might cause NPE which says nothing.
Does wrong username/password returns some generic spring "403 HTTP code exception"? 

Other option would be to use `exchange` method variant with ParameterizedTypeReference argument - that way you can specify what you expect (class) from the HTTP API and "lock" the response. I also guess, that way you will get more descriptive exception from Spring if the response is not as you would expect.

Currently, the user of this method needs to know there can be spring 403 if password is wrong; some other exception if server is offline/wrong URL has been provided; NPE when response is not as expected. At least, it could be wrapped by exception like: "Failed to obtain token from DHIS2 server at {url}, reason: {cause.getMessage()}."

Reply:
fixed, added exception classes and exception handling methods
I also added message keys class based on [referencedata MessageKeys](https://github.com/OpenLMIS/openlmis-referencedata/blob/master/src/main/java/org/openlmis/referencedata/util/messagekeys/MessageKeys.java)

Response statuses for methods handling exceptions were selected based on [HTTP Status Code for External Dependency Error](https://stackoverflow.com/questions/25398364/http-status-code-for-external-dependency-error) and https://stackoverflow.com/questions/18709834/http-status-424-or-500-for-error-on-external-dependency